README.md
<a href="#top" id="top">
  <img src="https://user-images.githubusercontent.com/441546/36892807-cca8da20-1dba-11e8-8a8c-8928daf3f779.png">
</a>
<p align="center">
  <a href="https://www.npmjs.com/package/@darkobits/doorman"><img src="https://img.shields.io/npm/v/@darkobits/doorman.svg?style=flat-square"></a>
  <a href="https://travis-ci.org/darkobits/doorman"><img src="https://img.shields.io/travis/darkobits/doorman.svg?style=flat-square"></a>
  <a href="https://www.codacy.com/app/darkobits/doorman"><img src="https://img.shields.io/codacy/coverage/1fbd5676e2df4ec78555c507a4d3d5a3.svg?style=flat-square"></a>
  <a href="https://github.com/conventional-changelog/standard-version"><img src="https://img.shields.io/badge/conventional%20commits-1.0.0-027dc6.svg?style=flat-square"></a>
  <a href="https://github.com/sindresorhus/xo"><img src="https://img.shields.io/badge/code_style-XO-e271a5.svg?style=flat-square"></a>
</p>


Modern apartment communities equipped with access control systems typically work like this:

1. The resident gives their phone number (usually a cell) to their apartment community.
2. When a guest arrives, they dial a number associated with the resident in the directory.
3. The access control system calls the resident.
4. The resident enters a digit or sequence of digits to open the door/gate for the guest.

This is annoying because of its synchronicity; it requires the resident to answer their phone, call quality is often very poor, and if anything goes wrong both parties wind up confused and frustrated. Additionally, this process does not scale well; a resident hosting a large event, for example, would have to remain tethered to their phone, anxious that one missed call will delay and frustrate one of their guests.

**What if there was _a better way_?**

1. The resident gives their phone number (in this case, a [Twilio](https://twilio.com) number that points to a server running Doorman) to their apartment community.
2. When a guest arrives, they dial a number associated with the resident in the directory.
3. The access control system calls Doorman which, depending on configuration, will:
   - Immediately dial the digit or sequence of digits to open the door/gate for the guest
   - Send the resident an SMS letting them know someone has arrived.

This process is completely asynchronous; the resident need not answer their phone and guests are never needlessly delayed or confused.

## Setup

Doorman is an HTTP server. To use it, you will need to install it someplace where HTTP servers like to live, such as [Heroku](https://heroku.com), [Now](https://zeit.co/now), [et. al](https://github.com/anaibol/awesome-serverless#hosting-and-code-execution-in-the-cloud).

Next, you will need to set up a Twilio account. Create a [TwiML App](https://support.twilio.com/hc/en-us/articles/223180928-How-do-I-create-a-TwiML-App-) and configure the Voice URL to point to the server running Doorman.

Doorman is configured using a key/value store where each key is a phone number and each value is a JSON blob describing how Doorman should behave. You can use any key/value store you like, including an in-memory object, but Redis is ideal, and Heroku offers free nodes.

### Installation

Doorman is available on NPM:

```bash
$ npm i @darkobits/doorman
```

### Configuration

Doorman has the following configurable options:

|Option|Description|
|---|---|
|`assetPath`|Path to a folder containing any static assets (such as audio files) you wish to serve.|
|`port`|Which port to listen on. This is typically configured automatically by providers like Heroku. Default: `8080`|
|`primaryPhoneNumber`|If Doorman can't find a matching key in its data store for an incoming call, it will forward the call to this number.|
|`twilioAccountSid`|Twilio Account SID.|
|`twilioApplicationSid`|Twilio Application SID.|
|`callDataFn`|This function will be invoked by Doorman and will be passed the current incoming caller ID and a callback. The callback has the signature `(err, data)` and should be invoked and passed the call flow data matching the provided caller ID.|
|`logLevel`|How much logging information to display. Uses [npmlog](https://github.com/npm/npmlog/blob/541407008c509755255a4819606e7916d26a77f5/log.js#L296-L304). (Default: `info`)|
|`allowInsecure`|Whether to allow connections over HTTP (default: `false`).|

### Examples

Here is an example using Redis as a data store:

```js
import { resolve } from 'path';
import redis from 'redis';
import doorman from '@darkobits/doorman';

const {
  PORT
  PRIMARY_PHONE_NUMBER,
  REDIS_URL,
  TWILIO_ACCOUNT_SID,
  TWILIO_APPLICATION_SID
} = process.env;


const client = new redis.createClient({
  url: REDIS_URL
});


doorman({
  assetPath: resolve(__dirname, 'assets'),
  port: PORT,
  primaryPhoneNumber: PRIMARY_PHONE_NUMBER,
  twilioAccountSid: TWILIO_ACCOUNT_SID,
  twilioApplicationSid: TWILIO_APPLICATION_SID,
  callDataFn: client.get.bind(client)
})
.startServer();
```

And here is an example using a basic in-memory data store:

```js
import { resolve } from 'path';
import doorman from '@darkobits/doorman';

const {
  PORT
  PRIMARY_PHONE_NUMBER,
  REDIS_URL,
  TWILIO_ACCOUNT_SID,
  TWILIO_APPLICATION_SID
} = process.env;

const data = {
  // ...
};


doorman({
  assetPath: resolve(__dirname, 'assets'),
  port: PORT,
  primaryPhoneNumber: PRIMARY_PHONE_NUMBER,
  twilioAccountSid: TWILIO_ACCOUNT_SID,
  twilioApplicationSid: TWILIO_APPLICATION_SID,
  callDataFn: (callerId, cb) => cb(undefined, data[callerId])
})
.startServer();
```

## Scripting Calls

Doorman is programmed using JSON. Each key in Doorman's data-store represents an inbound caller ID, and each value represents how Doorman should handle calls from that number. Doorman supports numerous directives which can be composed to construct a call flow.

The basic structure of a call flow is:

```js
[
  ["directiveName", { /* Directive options. */ }],
  ["directiveName", { /* Directive options. */ }],
  ["directiveName", { /* Directive options. */ }]
]
```

### Directives

#### `forwardCall`

Forwards the call to the provided number.

|Option|Type|Description|
|---|---|---|
|`value`|`String`|Number to forward to.|

**Example:**

```json
[
  ["forwardCall", {
    "value": "+14155551212"
  }]
]
```

#### `sendSms`

Sends a text message.

|Option|Type|Description|
|---|---|---|
|`value`|`String`|Body of the message.|
|`to`|`String`|Number to send the message to.|

**Example:**

```json
[
  ["sendSms", {
    "to": "+14155551212",
    "value": "Hello, world!"
  }]
]
```

#### `say`

Uses Twilio's text-to-speech feature to speak the provided message. See the [Twilio \<Say\> documentation](https://www.twilio.com/docs/api/twiml/say) for allowed values.

|Option|Type|Description|
|---|---|---|
|`value`|`String`|Message to speak.|
|`[voice='woman']`|`String`|Voice to use.|
|`[language='en-GB']`|`String`|Language to use.|

**Example:**

```json
[
  ["say", {
    "value": "Greetings!"
  }]
]
```

#### `sendDigits`

Sends a sequence of DTMF tones for the provided digit or digits.

|Option|Type|Description|
|---|---|---|
|`value`|`String`|Digit or digit sequence.|

**Example:**

```json
[
  ["sendDigits", {
    "value": "1234"
  }]
]
```

#### `gatherDigits`

Pauses the call and waits for the caller to enter a sequence of digits. The call will then proceed down the branch matching the sequence entered. A `default` branch must be provided to handle cases where the caller does not enter a matching sequence.

|Option|Type|Description|
|---|---|---|
|`branches`|`Object`|Object mapping possible responses to nested Doorman JSON blobs.

**Example:**

```js
[
  ["gatherDigits", {
    "123": [
      ["directiveName", { /* Directive options. */ }]
    ],
    "456": [
      ["directiveName", { /* Directive options. */ }]
    ],
    "default": [
      ["directiveName", { /* Directive options. */ }]
    ]
  }]
]
```

#### `play`

Instructs Twilio to play the audio file at the provided URL. To ensure Doorman serves static assets correctly, configure the `assetPath` option to point to the folder containing your audio files.

|Option|Type|Description|
|---|---|---|
|`value`|`String`|Path to the audio file (relative to Doorman's web root) to play.|

**Example:**

```json
[
  ["play", {
    "value": "assets/foo.mp3"
  }]
]
```

#### `hangUp`

Ends the call.

> **Note:** This directive is typically not needed, as Doorman will end the call if it reaches the end of the call flow.

**Example:**

```json
[
  ["hangUp"]
]
```

## Examples

In the following examples, let's assume we are working with an access control system that has the phone number `+14155551111`, a resident who has a cell number `+14155552222`, and that residents must enter the digit `9` to grant access to guests. Note that if using a data store such as Redis, we would store the access control system's caller ID as a key and the call flow as a value. These examples use a plain object for clarity.

In this first example, we will program Doorman to immediately grant access to guests, then send an SMS to the resident.

```json
{
  "+14155551111": [
    ["sendDigits", {
      "value": "9"
    }],
    ["sendSms", {
      "to": "+14155552222",
      "value": "A guest has arrived!"
    }]
  ]
}
```

Next, let's prompt the guest for a simple passcode (`123`) and forward the call to the resident if an incorrect passcode is entered:

```json
{
  "+14155551111": [
    ["say", {
      "value": "Please enter your passcode."
    }],
    ["gatherDigits", {
      "123": [
        ["say", {
          "value": "Access granted!"
        }],
        ["sendDigits", {
          "value": "9"
        }],
        ["sendSms", {
          "to": "+14155552222",
          "value": "A guest has arrived!"
        }]
      ],
      "default": [
        ["forwardCall", {
          "value": "+14155552222"
        }]
      ]
    }]
  ]
}
```

## &nbsp;
<p align="center">
  <br>
  <img width="22" height="22" src="https://cloud.githubusercontent.com/assets/441546/25318539/db2f4cf2-2845-11e7-8e10-ef97d91cd538.png">
</p>


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.2.0"></a>
# [0.2.0](https://github.com/darkobits/doorman/compare/v0.1.3...v0.2.0) (2018-03-02)


### Bug Fixes

* Correct allowInsecure logic. ([92f318d](https://github.com/darkobits/doorman/commit/92f318d))


### Features

* Add allowInsecure option. ([3887d6b](https://github.com/darkobits/doorman/commit/3887d6b))
* Primary Twilio number is no longer required. ([82f1dea](https://github.com/darkobits/doorman/commit/82f1dea))



<a name="0.1.3"></a>
## 0.1.3 (2017-09-07)


tests/helpers.js
import R from 'ramda';


/**
 * Asserts that the "expected" XML fragment is contained in the received XML
 * response from supertest.
 *
 * @param  {string} expected - XML fragment expected in the response.
 * @param  {object} response - Response from supertest.
 */
export const assertXmlContains = R.curry((expected, response) => {
  // Normalize both XML chunks by removing whitespace.
  expect(response.replace(/[\s]/g, '')).toMatch(expected.replace(/[\s]/g, ''));
});


tests/twiml.spec.js
import twimlResponse from '../src/lib/twiml';

import {
  DEFAULT_TIMEOUT,
  TWILIO_ENDPOINT
} from '../src/etc/constants';

import {
  assertXmlContains
} from './helpers';


const FROM_NUMBER = '415-555-1111';
const TO_NUMBER = '415-555-4567';


describe('TwiML Responses', () => {
  let twiml;

  beforeEach(() => {
    twiml = new twimlResponse(FROM_NUMBER, TO_NUMBER);
  });

  describe('forwardCall', () => {
    it('should create the correct TwiML', () => {
      const toNumber = '415-555-4444';

      twiml.forwardCall({
        value: toNumber
      });

      assertXmlContains(`
        <Dial action="${TWILIO_ENDPOINT}"
          callerId="${FROM_NUMBER}"
          method="GET"
          timeout="${DEFAULT_TIMEOUT}">
          ${toNumber}
        </Dial>
      `, twiml.render());
    });

    describe('not providing a value', () => {
      it('should throw an error', () => {
        expect(() => {
          twiml.forwardCall();
        }).toThrow();
      });
    });
  });

  describe('sendSms', () => {
    const toNumber = '415-555-4444';
    const message = 'foo';

    it('should create the correct TwiML', () => {
      twiml.sendSms({
        to: toNumber,
        value: message
      });

      assertXmlContains(`
        <Sms from="${TO_NUMBER}"
          to="${toNumber}">
          ${message}
        </Sms>
      `, twiml.render());
    });

    describe('not providing a value', () => {
      it('should throw an error', () => {
        expect(() => {
          twiml.sendSms({
            to: toNumber
          });
        }).toThrow('Did not receive a value');
      });
    });

    describe('not providing a "to" number', () => {
      it('should throw an error', () => {
        expect(() => {
          twiml.sendSms({
            value: message
          });
        }).toThrow('Did not receive a "to" number');
      });
    });
  });

  describe('say', () => {
    it('should create the correct TwiML', () => {
      const message = 'foo';

      twiml.say({
        value: message
      });

      assertXmlContains(`
        <Say language="en-GB"
          voice="woman">
          ${message}
        </Say>
      `, twiml.render());
    });

    describe('not providing a value', () => {
      it('should throw an error', () => {
        expect(() => {
          twiml.say();
        }).toThrow();
      });
    });
  });

  describe('sendDigits', () => {
    it('should create the correct TwiML', () => {
      const digits = '42';

      twiml.sendDigits({
        value: digits
      });

      assertXmlContains(`
        <Play digits="${digits}"></Play>
      `, twiml.render());
    });

    describe('not providing a value', () => {
      it('should throw an error', () => {
        expect(() => {
          twiml.sendDigits();
        }).toThrow();
      });
    });
  });

  describe('gatherDigits', () => {
    const digits = '42';

    it('should create the correct TwiML', () => {
      twiml.gatherDigits({
        [digits]: [],
        default: []
      });

      assertXmlContains(`
        <Gather action="${TWILIO_ENDPOINT}"
          method="GET"
          numDigits="${digits.length}"
          timeout="${DEFAULT_TIMEOUT}">
        </Gather>
      `, twiml.render());
    });

    describe('not providing a "default" branch', () => {
      it('should throw an error', () => {
        expect(() => {
          twiml.gatherDigits({
            [digits]: []
          });
        }).toThrow();
      });
    });
  });

  describe('play', () => {
    it('should create the correct TwiML', () => {
      const file = '/foo/bar.baz';

      twiml.play({
        value: file
      });

      assertXmlContains(`
        <Play>${file}</Play>
      `, twiml.render());
    });

    describe('not providing a value', () => {
      it('should throw an error', () => {
        expect(() => {
          twiml.play();
        }).toThrow();
      });
    });
  });

  describe('hangUp', () => {
    it('should create the correct TwiML', () => {
      twiml.hangUp();

      assertXmlContains(`
        <Pause length="1"></Pause>
        <Hangup></Hangup>
      `, twiml.render());
    });
  });
});


tests/server.spec.js
import R from 'ramda';
import req from 'supertest';
import Doorman from '../src';

import {
  TWILIO_ENDPOINT
} from '../src/etc/constants';

import {
  assertXmlContains
} from './helpers';


const FROM_NUMBER = '415-555-1111';
const PRIMARY_PHONE_NUMBER = '415-555-2222';
const TO_NUMBER = '415-555-4444';


/**
 * Creates a Doorman instance.
 *
 * @param  {object} data - Mock data to use.
 * @return {object}
 */
function createApp (data) {
  const TWILIO_ACCOUNT_SID = '1';
  const TWILIO_APPLICATION_SID = '2';

  const app = Doorman({
    logLevel: 'error',
    primaryPhoneNumber: PRIMARY_PHONE_NUMBER,
    twilioAccountSid: TWILIO_ACCOUNT_SID,
    twilioApplicationSid: TWILIO_APPLICATION_SID,
    callDataFn (callerId, cb) {
      if (data[callerId]) {
        cb(undefined, data[callerId]);
      } else {
        cb(undefined, null);
      }
    }
  }).server;


  /**
   * Simulates a request to Doorman coming from Twilio.
   *
   * @param  {object} params
   * @return {promise}
   */
  const request = params => req(app)
    .get(TWILIO_ENDPOINT)
    .set('Accept', 'application/xml')
    .set('X-Forwarded-Proto', 'https')
    .query(Object.assign({
      AccountSid: TWILIO_ACCOUNT_SID,
      ApplicationSid: TWILIO_APPLICATION_SID,
      CallSid: '1',
      From: FROM_NUMBER,
      To: TO_NUMBER
    }, params));


  return {
    request
  };
}


describe('Doorman', () => {
  let app;

  describe('Hello World', () => {
    const value = 'Hello world.';

    const call = [
      ['say', {value}]
    ];

    const database = {
      [FROM_NUMBER]: call
    };

    beforeAll(() => {
      app = createApp(database);
    });

    it('should say "Hello world."', () => {
      return app.request()
      .expect(200)
      .then(R.prop('text'))
      .then(assertXmlContains(`
        <Say language="en-GB"
          voice="woman">
          ${value}
        </Say>
        <Pause length="1"></Pause>
        <Hangup></Hangup>
      `));
    });
  });


  describe('Call Data Not Found', () => {
    beforeAll(() => {
      app = createApp({});
    });

    it('should forward the call to the primary phone number', () => {
      return app.request()
      .expect(200)
      .then(R.prop('text'))
      .then(assertXmlContains(`
        <Dial action="${TWILIO_ENDPOINT}"
          callerId="${FROM_NUMBER}"
          method="GET"
          timeout="20">
          ${PRIMARY_PHONE_NUMBER}
        </Dial>
      `));
    });
  });


  describe('Invalid requests', () => {
    it('should return a 400 error', () => {
      return app.request({
        ApplicationSid: false,
        AccountSid: false
      })
      .expect(400);
    });
  });


  describe('Send Digits & Hang Up', () => {
    const digits = '42';

    const call = [
      ['sendDigits', {
        value: digits
      }]
    ];

    const database = {
      [FROM_NUMBER]: call
    };

    beforeAll(() => {
      app = createApp(database);
    });

    it('should say "Hello world."', () => {
      return app.request({
        From: FROM_NUMBER
      })
      .expect(200)
      .then(R.prop('text'))
      .then(assertXmlContains(`
        <Play digits="${digits}"></Play>
        <Pause length="1"></Pause>
        <Hangup></Hangup>
      `));
    });
  });


  describe('Gather Digits', () => {
    const defaultResponse = 'foo';

    const sequence1 = '123';
    const response1 = `You entered ${sequence1}.`;

    const longSequence = '613731214614354234123';

    const call = [
      ['gatherDigits', {
        [sequence1]: [
          ['say', {
            value: response1
          }]
        ],
        [longSequence]: [],
        default: [
          ['say', {
            value: defaultResponse
          }]
        ]
      }]
    ];

    const database = {
      [FROM_NUMBER]: call
    };

    describe('Entering a matching sequence', () => {
      beforeAll(() => {
        app = createApp(database);
      });

      it('should send a gather instruction', () => {
        return app.request()
        .expect(200)
        .then(R.prop('text'))
        .then(assertXmlContains(`
          <Gather action="/twilio" method="GET" numDigits="${longSequence.length}" timeout="20"></Gather>
        `));
      });

      it('should follow the branch corresponding to the sequence entered', () => {
        return app.request({
          Digits: sequence1
        })
        .expect(200)
        .then(R.prop('text'))
        .then(assertXmlContains(`
          <Say language="en-GB" voice="woman">${response1}</Say>
          <Pause length="1"></Pause>
          <Hangup></Hangup>
        `));
      });
    });


    describe('Not entering a matching sequence', () => {
      beforeAll(() => {
        app = createApp(database);
      });

      it('should send a gather instruction', () => {
        return app.request()
        .expect(200)
        .then(R.prop('text'))
        .then(assertXmlContains(`
          <Gather action="/twilio"
            method="GET"
            numDigits="${longSequence.length}"
            timeout="20">
          </Gather>
        `));
      });

      it('should follow the branch corresponding to the sequence entered', () => {
        return app.request({
          Digits: '555'
        })
        .expect(200)
        .then(R.prop('text'))
        .then(assertXmlContains(`
          <Say language="en-GB"
            voice="woman">
            ${defaultResponse}
          </Say>
          <Pause length="1"></Pause>
          <Hangup></Hangup>
        `));
      });
    });
  });
});


src/lib/call.js
// ----- Call Parser -----------------------------------------------------------

import {is} from 'ramda';
import {DEFAULT_ACTION} from '../etc/constants';
import twimlResponse from './twiml';


/**
 * Creates a new call instance based on the provided call structure. Provides
 * methods to advance through the call based on user input.
 *
 * @param {object} params
 * @param {string} params.callId - Twilio Call SID.
 * @param {string} params.fromNumber - Inbound caller ID.
 * @param {array}  params.call - Call sequence.
 *
 * @return {object}
 */
export default function createCall ({call, callId, fromNumber, toNumber} = {}) {
  if (!call || !is(Array, call)) {
    throw new Error('createCall requires params.call be of type "Array".');
  }

  if (!callId) {
    throw new Error('createCall requires params.callId be set.');
  }

  if (!fromNumber) {
    throw new Error('createCall requires params.fromNumber be set.');
  }

  if (!toNumber) {
    throw new Error('createCall requires params.toNumber be set.');
  }


  /**
   * Index pointing to the current segment in the call structure.
   *
   * @type {number}
   */
  let curSegment = 0;


  /**
   * Whether the call is "paused" (waiting for user input) or not.
   *
   * @type {boolean}
   */
  let paused = false;


  /**
   * Whether the call is completed or not.
   *
   * @type {array}
   */
  let completed = false;


  /**
   * Reference to a callback that can be invoked to resume the call and set the
   * new call structure.
   *
   * @type {function}
   */
  let resumeCallback = null;


  /**
   * Reference to the current TwiML response instance.
   *
   * @type {object}
   */
  const twiml = twimlResponse(fromNumber, toNumber);


  /**
   * @private
   *
   * Advances the call until a terminal TwiML segment is reached. Returns the
   * compiled chunk of TwiML.
   *
   * @return {array} - List of TwiML chunks.
   */
  function getNextSegment () {
    // If the call is completed, it cannot be advanced.
    if (completed) {
      throw new Error('[getNextSegment] Trying to advance a completed call.');
    }

    let result;
    let terminal;
    const segment = call[curSegment];

    if (segment) {
      // Use the segment to generate TwiML.
      const [command, params] = segment;

      result = twiml[command](params);
      terminal = result.terminal;
    } else {
      // Otherwise, we have reached the end of the call or an invalid function
      // was provided in the call structure. Use the default action to generate
      // TwiML and set terminal and completed to true.
      result = twiml[DEFAULT_ACTION]();
      terminal = true;
      completed = true;
    }

    if (terminal) {
      // If we have a terminal chunk that has also provided a resume handler,
      // pause the call and set the resume callback.
      if (result.onResume) {
        paused = true;
        resumeCallback = result.onResume;
      } else {
        // If we have a terminal chunk without a resume handler, increment the
        // index.
        curSegment++;
      }

      // Because we have reached a terminal segment (ie: user input is required
      // to continue) render the current TwiML.
      const twimlDocument = twiml.render();

      // Clear all generated TwiML.
      twiml.reset();

      // Then, return the full TwiML document.
      return twimlDocument;
    }

    // Otherwise, if we have not reached a terminal segment, increment the
    // segment index and recurse to get the next TwiML chunk.
    curSegment++;

    return getNextSegment();
  }


  /**
   * @private
   *
   * Resumes the call by passing the provided arguments to the resume callback,
   * which will return a new call structure to parse.
   *
   * @param {object} resumeArgs - Argument to pass to resume callback.
   */
  function resume (resumeArgs) {
    if (!resumeCallback) {
      throw new Error('[resume] Call paused, but no resume handler was set.');
    }

    // Set the call structure to the return value of the resume callback.
    call = resumeCallback(resumeArgs);

    // Reset the segment index.
    curSegment = 0;

    // Unpause the call.
    paused = false;
  }


  /**
   * @return {boolean} - Whether the call is paused.
   */
  function isPaused () {
    return paused;
  }


  /**
   * @return {boolean} - Whether the call is completed.
   */
  function isCompleted () {
    return completed;
  }


  /**
   * Advances the call by optionally resuming it with the provided params and
   * then returning the next TwiML document.
   *
   * @param {object} - Params to pass to the resume callback.
   * @return {string} - Compiled TwiML chunk, if applicable.
   */
  function advance (params = {}) {
    if (paused) {
      resume(params);
    }

    // This will recurse until a terminal chunk or the end of a branch is
    // reached, ultimately returning a complete TwiML document.
    return getNextSegment();
  }


  return {
    advance,
    isCompleted,
    isPaused
  };
}


src/index.js
import 'babel-polyfill';

import path from 'path';
import bodyParser from 'body-parser';
import express from 'express';
import {equals, is, merge, where} from 'ramda';

import createCall from './lib/call';
import log from './lib/log';
import twimlResponse from './lib/twiml';

import {
  TWILIO_ENDPOINT
} from './etc/constants';


/**
 * Creates a new Doorman server. Accepts a config object with the following:
 *
 * @param {object} config - Config object.
 * @param {string} config.port - Port to listen on. If not provided, falls back
 *   to process.env.PORT, and then 8080.
 * @param {string} config.assetPath - Folder to serve static assets from.
 * @param {string} config.primaryPhoneNumber - Primary number for forwarding calls.
 * @param {string} config.twilioAccountSid - Twilio account SID.
 * @param {string} config.twilioApplicationSid - Twilio application SID.
 */
export default function Doorman (userConfig) {
  /**
   * Configuration for the instance.
   *
   * @type {object}
   */
  const config = Object.assign({
    assetPath: './assets',
    logLevel: 'verbose',
    port: 8080,
    allowInsecure: false
  }, userConfig);


  /**
   * Reference to the Express server.
   *
   * @type {object}
   */
  const server = express();


  /**
   * Tracks current calls.
   *
   * @type {map}
   */
  const calls = new Map();


  /**
   * Substitutes the Twilio web client's caller ID with the default Twilio
   * number so that call-forwarding from the Twilio web client works.
   *
   * @param {string} callerId - Phone number to parse.
   * @return {string} - Application's default phone number, or original caller ID.
   */
  function parseCallerId (callerId) {
    return callerId === 'client:Anonymous' ? config.twilioPhoneNumber : callerId;
  }


  /**
   * @private
   *
   * Express middleware used to ensure a request originated from Twilio.
   */
  function validateTwilioRequest (req, res, next) {
    if (process.env.NODE_ENV === 'development') {
      log.verbose('validateTwilioRequest', 'Not in production environment, assuming request is valid.');
      next();
      return;
    }

    const inboundCallerId = parseCallerId(req.query.From);
    const isSecure = config.allowInsecure || req.get('X-Forwarded-Proto') === 'https';

    log.verbose('validateTwilioRequest', `Incoming call from "${inboundCallerId}".`);

    const isValid = where({
      ApplicationSid: equals(config.twilioApplicationSid),
      AccountSid: equals(config.twilioAccountSid)
    }, merge(req.query, req.body));

    if (isSecure && isValid) {
      log.verbose('validateTwilioRequest', 'Twilio request is valid.');
      next();
    } else {
      log.warn('validateTwilioRequest', 'Twilio request is invalid.');
      res.sendStatus(400);
    }
  }


  /**
   * @private
   *
   * Defers to configured premise data function to fetch data, then validates
   * and returns it.
   *
   * @param  {string} callerId - Caller ID for the premise to get data for.
   * @return {Promise<Object>} - Resolves with premise data, or rejects on error.
   */
  function getCallData (callerId) {
    return new Promise((resolve, reject) => {
      config.callDataFn(callerId, (err, data) => {
        data = is(String, data) ? JSON.parse(data) : data;

        if (err) {
          reject(new Error(err));
          return;
        }

        if (data) {
          resolve(data);
        } else {
          reject(new Error(`No call data found for number "${callerId}".`));
        }
      });
    });
  }


  /**
   * Returns the current call from the calls Map if it exists. Otherwise,
   * creates a new call, adds it to the calls Map, and returns it.
   *
   * @param {string} callSid - Twilio Call SID.
   * @param {string} inboundCallerId - Number of the calling party.
   * @param {string} toNumber - Twilio number dialed by the calling party.
   */
  async function getOrCreateCall (callSid, inboundCallerId, toNumber) {
    if (calls.has(callSid)) {
      log.verbose('getOrCreateCall', `Call "${callSid}" exists, resuming call.`);
      return calls.get(callSid);
    }

    const callStruct = await getCallData(inboundCallerId);

    const call = createCall({
      callId: callSid,
      fromNumber: inboundCallerId,
      toNumber,
      call: callStruct
    });

    log.verbose('getOrCreateCall', `Call created for  "${callSid}".`);

    calls.set(callSid, call);

    return call;
  }


  /**
   * Starts the server.
   */
  function startServer () {
    server.listen(config.port, () => log.info('startServer', `Doorman running on port: ${config.port}`));
  }


  // ----- Initialization ------------------------------------------------------

  // Set the log level.
  log.level = config.logLevel;

  // Set up path for serving static assets.
  server.use(express.static(path.resolve(config.assetPath)));

  // Set up body parser middleware.
  server.use(bodyParser.urlencoded({
    extended: true
  }));

  // Endpoint used by Twilio.
  server.get(TWILIO_ENDPOINT, validateTwilioRequest, async (req, res) => {
    res.set('Content-Type', 'application/xml');

    const {CallSid, From, To} = req.query;
    const inboundCallerId = parseCallerId(From);

    try {
      // Get or create the instruction set for the call.
      const call = await getOrCreateCall(CallSid, inboundCallerId, To);

      // Get TwiML to send.
      const twiml = call.advance(req.query);

      // If the call is now completed, remove it from the calls map.
      if (call.isCompleted()) {
        log.verbose('request', `Call "${CallSid}" is complete.`);
        calls.delete(CallSid);
      }

      // Finally, send the TwiML.
      res.send(twiml);
    } catch (err) {
      log.warn('request', `Request failed: ${err}`);
      log.verbose('request', `Forwarding call to "${config.primaryPhoneNumber}".`);
      // Call data could not be found for the incoming caller, or an unknown
      // error occurred. Forward the call to the primary number.
      const twiml = twimlResponse(inboundCallerId, To);

      twiml.forwardCall({
        value: config.primaryPhoneNumber
      });

      res.send(twiml.render());
    }
  });


  return {
    server,
    startServer
  };
}
